### PR TITLE
Improvement/2016 brand the dex gui to meet the design

### PR DESCRIPTION
**Pull request by ChengYanJin**
**Component**: Dex, UI

**Context**: 
As I user, I want to have the same user experience as Metalk8s UI when I log in.

**Summary**:
We create a configMap and mount it as volume in the web/theme/scality
```
#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: dex-login
  namespace: metalk8s-auth
data:
  styles.css: @@styles
binaryData:
  favicon.png: @@favicon
  logo.png: @@branding
```
**Acceptance criteria**: 
1. ./doit.sh vagrant_up
2. Get the control plane ip
3.  Go to the login page
```https://<control_plane_ip>:8443/oidc/auth?audience=&amp;client_id=metalk8s-ui&amp;redirect_uri=https%3A%2F%2F<control_plane_ip>%3A8443%2Foauth2%2Fcallback&amp;response_type=id_token&amp;scope=openid+profile+email+offline_access&amp;nonce=abcde```
image

Closes: #2016

### CODE REVIEW

**Comment by gdemonet** [comment]
@ChengYanJin you'll need to rebase on `development/2.5` once #2063 gets merged (and probably will need the code from #2025 as well)

**Review by slaperche-scality — COMMENTED** [review]
LGTM for the buildchain part.
I tested it and it works.

Good job :+1:

File: buildchain/buildchain/salt_tree.py [hunk 1]
```diff
@@ -153,6 +153,25 @@ def _get_parts(self) -> Iterator[str]:
...
+
+
+def salt_embed_bytes(path: Path, indent: int) -> str:
+    """Return the content of a binary file in usable form by Salt."""
+    data = ['|']
+    bytestring = path.read_bytes()
+    b64data = base64.encodebytes(bytestring).decode('utf-8')
+    for line in b64data.split('\n'):
+        data.append('{}{}'.format(' '*indent, line.rstrip()))
+    return '\n'.join(data)
```
Comment (gdemonet) [inline buildchain/buildchain/salt_tree.py:172]: Mmpf, that's a strange way to do things. Both could rely on `yaml.dump` (see https://github.com/scality/metalk8s/pull/2042/commits/1ada06c5769e987afc7473cb58250419e413b5bd for an example of how to customize representers). But let's keep it for a debt ticket.

File: charts/dex.yaml [hunk 2]
```diff
@@ -52,8 +52,8 @@ config:
     tlsCert: /etc/dex/tls/https/server/tls.crt
     tlsKey: /etc/dex/tls/https/server/tls.key
   frontend:
-    theme: "coreos" #metalk8s-ui
-    # dir: /web/themes/custom/
+    theme: "scality"
+    issuer: "MetalK8s"
```
Comment (gdemonet) [inline charts/dex.yaml:56]: Must have took you quite some time to find out how this works, as it's not documented at all... I had to read the Go sources to understand what it does... Maybe worth a PR upstream to clarify this usage?

File: salt/metalk8s/addons/dex/deployed/login.sls.in [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
+
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: dex-login
```
Comment (gdemonet) [inline salt/metalk8s/addons/dex/deployed/login.sls.in]: Let's call this file `theme-configmap.sls` for clarity, or just `configmap` if you prefer. Login is too vague IMHO.

File: salt/metalk8s/addons/dex/deployed/login.sls.in [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
Comment (gdemonet) [inline salt/metalk8s/addons/dex/deployed/login.sls.in]: After #2050 we'll stop using `kubeconfig` and `context` completely, so I don't know how you want to handle that, maybe rebase and remove it right now...

**Review by gdemonet — CHANGES_REQUESTED** [review]
Minor changes, good job! We'll need to think about a different approach though (I guess fetch the assets from MetalUI directly could work, so that if the theme changes, this one will as well)

File: salt/metalk8s/addons/dex/deployed/login.sls.in (hunk 4 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/addons/dex/deployed/login.sls.in]: We still support it just it's not needed so we can merge like this and remove it after

**Comment by ChengYanJin** [comment]
> Minor changes, good job! We'll need to think about a different approach though (I guess fetch the assets from MetalUI directly could work, so that if the theme changes, this one will as well)

Hi Guillaume, thanks for the review!

Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
The reason is that I need to follow the default file name of Dex.

I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

**Comment by gdemonet** [comment]
> Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
> The reason is that I need to follow the default file name of Dex.

Oh OK. Actually my point was that, instead of mounting those assets in the Dex container, we could find some way to let them be referenced by URL instead (such URL pointing to Metal UI). But really not urgent. I think I've seen a `logoURL` configuration option in some examples, for instance.

> I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

Yeah it will make sense if we manage to serve them with Metal UI (and not store them in a ConfigMap at build time). So let's keep it, good idea :)

**Comment by gdemonet** [comment]
> Actually I fetch different asset for Dexlogin and Metalk8sUI. e.g For scality logo, we use branding.svg => MetalK8s and logo.png => Dexlogin. 
> The reason is that I need to follow the default file name of Dex.

Oh OK. Actually my point was that, instead of mounting those assets in the Dex container, we could find some way to let them be referenced by URL instead (such URL pointing to Metal UI). But really not urgent. I think I've seen a `logoURL` configuration option in some examples, for instance.

> I moved the dexlogin assets in the ui/public/brand/assets/login. Wondering does it make sense to you?

Yeah it will make sense if we manage to serve them with Metal UI (and not store them in a ConfigMap at build time). So let's keep it, good idea :)

**Review by gdemonet — APPROVED** [review]

File: charts/dex.yaml [hunk 5]
```diff
@@ -31,14 +31,14 @@ ingress:
...
     - null
 
-# extraVolumes:
-#   - name: theme
-#     configMap:
-#        name: dex-branding
+extraVolumes:
+  - name: dex-login
+    configMap:
+      name: dex-login
```
Comment (NicolasT) [inline charts/dex.yaml:37]: I wish we'd have given this a less generic name like `dex-scality-theme` or whatever.

Comment (ChengYanJin) [reply charts/dex.yaml:37]: > I wish we'd have given this a less generic name like `dex-scality-theme` or whatever.

I will rename it in the 2nd iteration of authentication.

**Comment by NicolasT** [comment]
Don't postpone renaming of things. It's a hassle because of extra
upgrade/downgrade automation required.

**Pull request merged**